lib/chef/userable.rb

```diff
@@ -0,0 +1,106 @@
+#
+# Author:: Snehal Dwivedi (sdwivedi@chef.io)
+# Copyright:: Copyright (c) 2008-2016 Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# common methods of User and UserV1
+class Chef
+  class Userable
+
+    def initialize
+      @name = ""
+      @password = nil
+      @admin = false
+      @username = nil
+      @display_name = nil
+      @first_name = nil
+      @middle_name = nil
+      @last_name = nil
+      @email = nil
+      @public_key = nil
+      @private_key = nil
+      @create_key = nil
+    end
+
+    def chef_root_rest_v0
+      @chef_root_rest_v22 ||= Chef::ServerAPI.new(Chef::Config[:chef_server_root], { api_version: "0" })
+    end
+
+    def chef_root_rest_v1
+      @chef_root_rest_v22 ||= Chef::ServerAPI.new(Chef::Config[:chef_server_root], { api_version: "1" })
+    end
```

Comment:
We shouldn't need explicit v1 here, the Userv1 API is compatible with the v2 API as well and we should be able to use the same rest object that all the other APIs consume, and not have to recreate a new one.

The naming of the instance variable is also a bit odd

Reply:
For memoization, the ruby convention is to name the instance variable the same as the method. It should definitely not get reused between methods.